pkg/server/commands/reverseexpand/reverse_expand.go

```diff
@@ -267,39 +258,29 @@ func (c *ReverseExpandQuery) execute(
 
 	for _, edge := range edges {
 		innerLoopEdge := edge
-		if currentEdge != nil && currentEdge.Condition == graph.RequiresFurtherEvalCondition {
-			// propagate the condition to upcoming reverse expansions
-			// TODO don't mutate the edge, keep track of the previous edge's condition and use it in trySendCandidate
-			innerLoopEdge.Condition = graph.RequiresFurtherEvalCondition
-		}
+		intersectionOrExclusionInPreviousEdges := intersectionOrExclusionInPreviousEdges || innerLoopEdge.TargetReferenceInvolvesIntersectionOrExclusion
 		subg.Go(func() error {
-			r := &reverseExpandRequest{
-				storeID:          storeID,
+			r := &ReverseExpandRequest{
+				StoreID:          req.StoreID,
+				ObjectType:       req.ObjectType,
+				Relation:         req.Relation,
+				User:             req.User,
+				ContextualTuples: req.ContextualTuples,
 				edge:             innerLoopEdge,
-				targetObjectRef:  targetObjRef,
-				sourceUserRef:    req.User,
-				contextualTuples: req.ContextualTuples,
 			}
 
 			switch innerLoopEdge.Type {
-			case graph.DirectEdge:
-				return c.reverseExpandDirect(subgctx, r, resultChan, resolutionMetadata)
+			case graph.DirectEdge, graph.TupleToUsersetEdge:
+				return c.readTuplesAndExecute(subgctx, r, resultChan, intersectionOrExclusionInPreviousEdges, resolutionMetadata)
```

Comment:
Though this reduces some duplicate code paths, here are some downsides that I'm not particularly in favor of:

1. We lose trace clarity.

   Before this it was clear the particular resolution path we were taking by setting a trace named `reverseExpandDirect` or `reverseExpandTTU`. Now those independent paths are obfuscated.
   
   2. We introduce variations in evaluation similarities between Check and ListObjects related code.
   
       In Check resolution we have specific handlers for each terminal rewrite path we can encounter `checkDirect`, `checkComputedUserset`, `checkTTU`. The code before these changes mimicked that as well with `reverseExpandDirect`, `reverseExpandComputedUserset`, `reverseExpandTTU`. 
       
I would recommend we de-dup the shared code as a common helper, but reintroduce the prior composition for the aforementioned reasons above.

Reply:
> Before this it was clear the particular resolution path we were taking by setting a trace named reverseExpandDirect or reverseExpandTTU. Now those independent paths are obfuscated.

The span `readTuplesAndExecute` contains the attribute `edge` which will tell if you if the edge is a Direct or a TTU

> We introduce variations in evaluation similarities between Check and ListObjects related code.

🤔 there are no similarities? ListObjects is doing BFS of a graph. Check isn't using the graph at all. The only similarities would be in the name of the functions 🤷‍♀️ 

My concern is that if we don't remove the duplicate code, when we come back to change something about the reads (for example), we may forget to change the other function too.

Reply:
@jon-whit, in the last commit I kept the original functions 👍